x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/apikey/ApiKeySingleNodeTests.java

```diff
@@ -238,6 +255,133 @@ public void testGetApiKeyWorksForTheApiKeyItself() {
         assertThat(e.getMessage(), containsString("unauthorized for API key id [" + apiKeyId + "]"));
     }
 
+    public void testGrantApiKeyForUserWithRunAs() throws IOException {
+        final TestSecurityClient securityClient = getSecurityClient();
+        securityClient.putRole(new RoleDescriptor("user1_role", new String[] { "manage_token" }, null, new String[] { "user2", "user4" }));
+        securityClient.putRole(new RoleDescriptor("user2_role", new String[] { "monitor" }, null, null));
+        final SecureString user1Password = new SecureString("user1-strong-password".toCharArray());
+        securityClient.putUser(new User("user1", "user1_role"), user1Password);
+        securityClient.putUser(new User("user2", "user2_role"), new SecureString("user2-strong-password".toCharArray()));
+        securityClient.putUser(new User("user3", "user3_role"), new SecureString("user3-strong-password".toCharArray()));
+
+        // Success: user1 runas user2
+        final GrantApiKeyRequest grantApiKeyRequest = buildGrantApiKeyRequest("user1", user1Password, "user2", securityClient);
+        final CreateApiKeyResponse createApiKeyResponse = client().execute(GrantApiKeyAction.INSTANCE, grantApiKeyRequest).actionGet();
+        final String apiKeyId = createApiKeyResponse.getId();
+        final String base64ApiKeyKeyValue = Base64.getEncoder()
+            .encodeToString((apiKeyId + ":" + createApiKeyResponse.getKey().toString()).getBytes(StandardCharsets.UTF_8));
+        assertThat(securityClient.getApiKey(apiKeyId).getUsername(), equalTo("user2"));
+        // The API key has privileges (inherited from user2) to check cluster health
+        client().filterWithHeader(Map.of("Authorization", "ApiKey " + base64ApiKeyKeyValue))
+            .execute(ClusterHealthAction.INSTANCE, new ClusterHealthRequest())
+            .actionGet();
```

Comment:
Can we have a negative test here too?
That is, give `user1_role` a privilege that is easy to test for. Don't give that privilege to `user2_role`. Check that the API Key doesn't have the privilege.

One of the possible bugs that could exist (or be introduced in the future) here is that the API key has the `monitor` privilege due to the facilitating user (`test_user`) which has cluster `ALL`.
Testing that the API Key _doesn't_ have something it isn't supposed to have helps minimize that risk.

Reply:
And can we also have a test with limiting roles? When there's 3 sets of users involved, there's a risk that something causes us to pull in the wrong set of roles, and I'd like to test more extensively.

Reply:
> So I assume you mean the API key should have its own role descriptors?

Yes, that's what I meant. Did I use the wrong words? I can never remember which one we call `limiting` and `limited`.